### PR TITLE
feat : add secure feature

### PR DESCRIPTION
**Pull request by YUandJIANG**
#### What type of PR is this?

feat
#### What this PR does / why we need it (en: English/zh: Chinese):

zh: 给 hertz 添加 secure 中间件

### CODE REVIEW

**Comment by welkeyever** [comment]
Please use the `README.md` to do the introduction and usage like this: https://github.com/hertz-contrib/swagger

File: README.md [hunk 1]
```diff
@@ -1 +1,125 @@
-# .github
\ No newline at end of file
+# Secure (This is a community driven project)
+
+`Secure` middleware for hertz framework.
+
+This repo is forked from [sessions](https://github.com/gin-contrib/secure) and adapted for hertz.
```
Comment (Skyenought) [inline README.md]: sessions -> secure

File: README.md [hunk 2]
```diff
@@ -1 +1,125 @@
...
+SSLProxyHeaders:       map[string]string{"X-Forwarded-Proto": "https"},
+}
+}
+```
+
+## Option
+
+| Option                | Default Value                                   | Description                                                  |
+| --------------------- | ----------------------------------------------- | ------------------------------------------------------------ |
+| SSLRedirect           | true                                            | If `SSLRedirect` is set to true, then only allow https requests |
```
Comment (Skyenought) [inline README.md]: 好像少了几个属性，推荐把 Default Value 改成 property

File: README.md (hunk 1 above)
Comment (YUandJIANG) [reply README.md]: GET

File: README.md (hunk 2 above)
Comment (YUandJIANG) [reply README.md]: GET

**Review by Skyenought — APPROVED** [review]
LGTM

File: README.md [hunk 3]
```diff
@@ -1 +1,133 @@
...
+#### Function Signature
+
+```go
+func DefaultConfig() Config
+```
+
+#### Sample Code
+
+```go
+package main
```
Comment (li-jin-gou) [inline README.md:26]: code need to format and readme need to improve

File: README.md [hunk 4]
```diff
@@ -1 +1,133 @@
...
+
+```bash
+go get github.com/hertz-contrib/secure
+```
+
+### [Default example](example/default/main.go)
+
+Default configuration for users to set security configuration directly using secure middleware
+
+#### Function Signature
```
Comment (li-jin-gou) [inline README.md:17]: Why do we need to introduce this function signature separately?

Comment (YUandJIANG) [reply README.md:17]: GET

File: README.md (hunk 3 above)
Comment (YUandJIANG) [reply README.md:26]: GET

File: secure.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+package secure
+
+import (
+	"context"
+
+	"github.com/cloudwego/hertz/pkg/app"
+)
+
+// Config is a struct for specifying configuration options for the secure.
+type Config struct {
```
Comment (li-jin-gou) [inline secure.go]: use `option` instead `config` refer to https://github.com/hertz-contrib/keyauth/blob/main/option.go

Comment (YUandJIANG) [reply secure.go]: GET

**Review by Skyenought — APPROVED** [review]
LGTM

File: policy.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	if cfg.browserXssFilter {
+		p.addHeader("X-Xss-Protection", "1; mode=block")
+	}
+
+	// Content Security Policy header.
+	if len(cfg.contentSecurityPolicy) > 0 {
+		p.addHeader("Content-Security-Policy", cfg.contentSecurityPolicy)
+	}
+
+	if len(cfg.referrerPolicy) > 0 {
```
Comment (li-jin-gou) [inline policy.go]: 为什么这里没有注释

File: secure.go [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+//		IsDevelopment:         false
+//		STSSeconds:            315360000
+//		STSIncludeSubdomains:  true
+//		FrameDeny:             true
+//		ContentTypeNosniff:    true
+//		BrowserXssFilter:      true
+//		ContentSecurityPolicy: "default-src 'self'"
+//		SSLProxyHeaders:       map[string]string{"X-Forwarded-Proto": "https"},
+// ```
+func Default(opts ...Option) app.HandlerFunc {
```
Comment (li-jin-gou) [inline secure.go]: 这里不是希望有这样一个 dfault的初始化方式，而是说option的默认值

Comment (li-jin-gou) [reply secure.go]: use option instead config refer to https://github.com/hertz-contrib/keyauth/blob/main/option.go

**Comment by li-jin-gou** [comment]
CI need to solve

File: .github/workflows/tests.yml [hunk 8]
```diff
@@ -17,13 +17,6 @@ jobs:
         with:
           go-version: ${{ matrix.go }}
 
-       # block scenario, comment temporarily
-#      - uses: actions/cache@v3
```
Comment (li-jin-gou) [inline .github/workflows/tests.yml:21]: 修改CI的原因是,这个模版生成的不需要修改的

File: .golangci.yaml [hunk 9]
```diff
@@ -1,36 +0,0 @@
-# Options for analysis running.
```
Comment (li-jin-gou) [inline .golangci.yaml]: +linter配置文件

File: README.md [hunk 10]
```diff
@@ -1 +1,94 @@
...
+	    browserXssFilter:      true,
+	    contentSecurityPolicy: "default-src 'self'",
+	    ieNoOpen:              true,
+	    sslProxyHeaders:       map[string]string{"X-Forwarded-Proto": "https"},
+       },
+```
+
+## Option
+
+| options                       | Parameters        | Description                                                  |
```
Comment (zstone12) [inline README.md:71]: 默认值可以单起一列，当然目前这样写在Description 也没啥问题:D 。

File: README.md [hunk 11]
```diff
@@ -1 +1,94 @@
...
+| WithSTSSecond                 | int64             | Default is 315360000, which would NOT include the header.    |
+| WithSTSIncludeSubdomains      | bool              | If `WithSTSIncludeSubdomains` is set to true, the `includeSubdomains` will be appended to the Strict-Transport-Security header. Default is false. |
+| WithFrameDeny                 | bool              | If `WithFrameDeny` is set to true, adds the X-Frame-Options header with the value of `DENY`. Default is false |
+| WithContentTypeNosniff        | bool              | If `WithContentTypeNosniff` is true, adds the X-Content-Type-Options header with the value `nosniff`. Default is false. |
+| WithBrowserXssFilter          | bool              | If `WithBrowserXssFilter` is true, adds the X-XSS-Protection header with the value `1; mode=block`. Default is false. |
+| WithContentSecurityPolicy     | []string          | `WithContentSecurityPolicy` allows the Content-Security-Policy header value to be set with a custom value. Default is "". |
+| WithIENoOpen                  | bool              | Prevent Internet Explorer from executing downloads in your site’s context |
```
Comment (zstone12) [inline README.md]: Prevent Internet Explorer from executing downloads in your site’s context 少一个.

File: README.md [hunk 12]
```diff
@@ -1 +1,94 @@
...
+	    ieNoOpen:              true,
+	    sslProxyHeaders:       map[string]string{"X-Forwarded-Proto": "https"},
+       },
+```
+
+## Option
+
+| options                       | Parameters        | Description                                                  |
+| ----------------------------- | ----------------- | ------------------------------------------------------------ |
+| WithSSLRedirect               | bool              | If `WithSSLRedirect` is set to true, then only allow https requests |
```
Comment (zstone12) [inline README.md]: `WithSSLRedirect` is set to true, then only allow https requests 同样少一个 .

File: .github/workflows/pr-check.yml [hunk 13]
```diff
@@ -54,7 +54,6 @@ jobs:
           go-version: 1.19
 
       - name: Golangci Lint
-        # https://golangci-lint.run/
-        uses: golangci/golangci-lint-action@v3
+        use: golangci/golangci-lint-action@v3
```
Comment (li-jin-gou) [inline .github/workflows/pr-check.yml]: gofumt 和 golangci-lint-action 的功能是重复的以及没有 go vet 检查
先将 
- pr-check.yml
- tests.yml
- .golangci.yaml 

复原吧

File: README.md [hunk 14]
```diff
@@ -1 +1,94 @@
...
+| WithCustomFrameOptionsValue   | string            | nil                         | `WithCustomFrameOptionsValue` allows the X-Frame-Options header value to be set with a custom value. This overrides the FrameDeny option. |
+| WithReferrerPolicy            | string            | nil                         | HTTP header "Referrer-Policy" governs which referrer information, sent in the Referrer header, should be included with requests made. |
+| WithBadHostHandler            | app.HandlerFunc   | nil                         | Handlers for when an error occurs (ie bad host).             |
+| WithFeaturePolicy             | string            | nil                         | Feature Policy is a new header that allows a site to control which features and APIs can be used in the browser. |
+| WithDontRedirectIPV4Hostnames | bool              | nil                         | If `WithDontRedirectIPV4Hostnames` is true, requests to hostnames that are IPV4 addresses aren't redirected. This is to allow load balancer health checks  to succeed. |
```
Comment (zstone12) [inline README.md]: | WithDontRedirectIPV4Hostnames | bool | nil `nil` 应该是typo?

File: .github/workflows/pr-check.yml (hunk 13 above)
Comment (YUandJIANG) [reply .github/workflows/pr-check.yml]: GET

**Comment by zstone12** [comment]
LGTM

**Review by zstone12 — APPROVED** [review]

**Comment by li-jin-gou** [comment]
CI 有问题，等待复原

**Comment by li-jin-gou** [comment]
LGTM

**Pull request merged**